File: 전유림/5주차[그래프]/24479.py
from collections import deque
import sys
sys.setrecursionlimit(10**9)
input = sys.stdin.readline

n, m, r = map(int, input().split())
graph = [[] for _ in range(n+1)]    # 각 노드가 연결된 정보?
visited = [0]*(n+1)
order = 1

for _ in range(m):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

def dfs(r):   # 정점, 간선, 시작
    global order
    visited[r] = order
    graph[r].sort()
    for i in graph[r]:
        if visited[i]==0:
            order += 1
            dfs(i)


dfs(r)

for i in visited[1:]:
    print(i)

File: sua/java/boj/graph/_24479.java
package boj.graph;

import java.io.*;
import java.util.*;

public class _24479 {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static int[] visited;
    static int trail;

    public static void main(String[] args) throws IOException{
        
        st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int R = Integer.parseInt(st.nextToken());

        for(int i = 0; i < N + 1; i++){
            graph.add(new ArrayList<>());
        }

        for(int m = 0; m < M; m++){

            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            graph.get(a).add(b);
            graph.get(b).add(a);

        }

        // 오름차순
        for(int i = 1; i < graph.size(); i++) {
            Collections.sort(graph.get(i));
        }

        visited = new int[N+1];
        trail = 1;
        dfs(R);

        for(int i = 1; i < visited.length; i++){
            sb.append(visited[i]).append("\n");
        }

        System.out.println(sb);
    }

    private static void dfs(int node) {
            visited[node] = trail++;

                for(int i = 0; i < graph.get(node).size(); i++){
                    if(visited[graph.get(node).get(i)] == 0) {
                        dfs(graph.get(node).get(i));
                    }
                }
            }
        
}
